File: Data Structures & Algorithms/search-for-word/submission-14.java
class Solution {
    int ROWS,COLS;
    boolean[][]vis;
    public boolean exist(char[][] board, String word) {
        ROWS = board.length;
        COLS = board[0].length;
        vis = new boolean[ROWS][COLS];
        for(int i=0;i<ROWS;i++){
            for(int j=0;j<COLS;j++){
                if(dfs(i,j,0,word,board)){
                    return true;
                }
            }
        }
        return false;

    }

    public boolean dfs(int i, int j, int k, String word, char[][] board){
        if (k == word.length())
            return true;
        if(i<0 || j<0|| i==ROWS || j == COLS|| word.charAt(k) != board[i][j] || vis[i][j]){
            return false;
        }
        
        vis[i][j] = true;
        if (dfs(i+1,j,k+1,word,board) || dfs(i,j+1,k+1,word,board) || 
        dfs(i-1,j,k+1,word,board) || dfs(i,j-1,k+1,word,board))
            return true;
        vis[i][j] = false;
        return false;
    }
}

File: Data Structures & Algorithms/regular-expression-matching/submission-14.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        dp = {}
        M,N = len(s), len(p)
        def dfs(i,j):
            print(i,j)
            if i == M and j == N:
                return True
            if j ==N:
                return False
            if (i,j) in dp:
                return dp[(i,j)]
            isMatch = i<M and (s[i] == p[j] or p[j] == ".")
            if j+1<N and p[j+1] == "*":
                dp[(i,j)] = (isMatch and dfs(i+1,j)) or dfs(i,j+2)
            elif isMatch:
                dp[(i,j)] = dfs(i+1,j+1)
            else:
                dp[(i,j)] = False
            return dp[(i,j)]
        return dfs(0,0)




        
        # i.     j 
        # abcd   .*

        # if i == len(s) and j == len(p):
        #     return True
        
        # if j+1 == "*":
            
        
        # elif isMatch
        